### DISCUSSION TITLE
Possible to make choice of database dynamic

### DISCUSSION THREAD

**Discussion by ErikEJ** [marcominerva/DatabaseGPT#68] [General]
Is it??

**Comment by marcominerva** [comment]
At this moment, the database is set at startup, but it is possible to change the service registration so that the configuration is dynamic. Let me make some test about this feature...

**Reply by ErikEJ** [reply]
That would be amazing, I have some ideas for some VS tooling based on this.

**Reply by ErikEJ** [reply]
Looks like entire settings object should be dynamic in this use case, so user can supply their own OpenAI key

**Reply by marcominerva** [reply]
For your use case, having the possibility to define the registration with a Service Provider that allows to read configuration for each request is enough (as the registration will be scoped)? For example:

```csharp
builder.Services.AddDatabaseGpt((provider, database) =>
{
    var userSettingsService = provider.GetRequiredService<IUserSettingsService>();
    var connectionString = userSettingsService.GetConnectionString();

    database.UseConfiguration(builder.Configuration).UseSqlServer(connectionString);
},
(provider, chatGpt) =>
{
    var userSettingsService = provider.GetRequiredService<IUserSettingsService>();
    var openAIApiKey = userSettingsService.GetOpenAIKey();

    chatGpt.UseOpenAI(openAIApiKey);
});
```

Or do you need to explicitly create a `DatabaseGptClient`?

**Reply by ErikEJ** [reply]
I think this should be fine, then I can implement my own UserSettingsService

I assume I can vary this based on the database type?

```csharp
builder.Services.AddDatabaseGpt((provider, database) =>
{
    var userSettingsService = provider.GetRequiredService<IUserSettingsService>();
    var connectionString = userSettingsService.GetConnectionString();

    database.UseConfiguration(builder.Configuration).UseSqlServer(connectionString);
},

```

**Reply by marcominerva** [reply]
Of course you can :)

I have just updated the `main` branch with this change. Take a look and tell me if it is OK for your need.

Note that `DatabaseGptClient` mantains a cache history for the conversation, for example to store the system message with all table names and custom information about the schema. If you change the database dynamically, I think you need to provide a new `sessionId `for each invocation of the `ExecuteNaturalLanguageQueryAsync` method.

We can also think about maintaining a separate cache history for each _(sessionId, connection string)_ couple, if it is worth the case for you.

**Reply by ErikEJ** [reply]
I actually plan to call my own .NET 8 console app on each "question", so this should be fine (I have to do that, as VS is a .NET Framework app)
